ramips: add support for Hongdian H8922 v30

rbqvq opened the pull request:
This is an industrial 4G router equipped with OpenWrt 14.07 OEM customized version

WARNING: The original firmware device tree is common to multiple boards, and the device tree name is H9350. This submitted device tree is a modified version, which deletes the non-this-device parts and adds GPIO watchdog.

Specification:
- SoC: MediaTek MT7620A
- Flash: 16 MB
- RAM: 128 MB
- Power: DC 5V-36V 1.5A
- Ethernet: 1x WAN, 4x LAN (10/100 Mbps)
- Wireless radio: 802.11n 2.4g-only
- LED: 
  - System/Power (RUN): GPIO/26 active-low
  - Ethernet: 1x WAN, 4x LAN
  - Modem 1: GPIO/66 active-low
  - RF 1 (Modem 1 Signal): GPIO/67 active-low
  - Modem 2: GPIO 71 active-low
  - RF 2 (Modem 2 Signal): GPIO/24 active-low 
  - WLAN: GPIO/72 active-low
  - WPS: GPIO/12 active-low
- Button: 
  - WPS / RESET: GPIO/34 active-low
- UART: 1x UART on PCB - 115200 8N1
- GPIO Watchdog: GPIO/62 mode=toggle timeout=1s
- PCIe: 2x miniPCIe for modem
- SIM Slots: 2x SIM Slots

Issue:
- No factory partition, eeprom is located at /lib/firmware/mt7620a.eeprom

Flash instruction:
Using UART:
1. Configure PC with a static IP address and setup an TFTP server.
2. Put rootfs into the tftp directory.
3. Connect the UART line as described on the PCB.
4. Power up the device and press Ctrl+C to break auto boot.
5. Use `system 6` command and follow the instruction to set device and tftp server IP address and input the rootfs file name. U-boot will then load the rootfs and write it into the flash.
6. Use `system 1` command and follow the instruction to set device and tftp server IP address and input the firmware file name. U-boot will then load the firmware once.
7. Login to LuCI and use LuCI upgrade firmware.

Original Firmware Dump / More details:
https://blog.gov.cooking/archives/research-hongdian-h8922-and-flash.html

---

Original U-Boot WebUI and UART Console upgrade firmware has header magic verification. We do not know it now.
But boot linux once, update kernel and update rootfs also check header. But its uImage and squashfs header.
Unfortunately, the kernel accepts a maximum of 1.5M.
So we must update rootfs first, and boot firmware once. (Directly boot initramfs-kernel was failed, it will auto reboot)
Then we can go to openwrt and flash full firmware to nor.

rbqvq committed:
ramips: add support for Hongdian H8922 v30

This is an industrial 4G router equipped with OpenWrt 14.07 OEM
customized version

WARNING: The original firmware device tree is common to multiple
boards, and the device tree name is H9350. This submitted device
tree is a modified version, which deletes the non-this-device parts
and adds GPIO watchdog.

Specification:
- SoC: MediaTek MT7620A
- Flash: 16 MB
- RAM: 128 MB
- Power: DC 5V-36V 1.5A
- Ethernet: 1x WAN, 4x LAN (10/100 Mbps)
- Wireless radio: 802.11n 2.4g-only
- LED:
  System/Power (RUN): GPIO/26 active-low
  Ethernet: 1x WAN, 4x LAN
  Modem 1: GPIO/66 active-low
  RF 1 (Modem 1 Signal): GPIO/67 active-low
  Modem 2: GPIO 71 active-low
  RF 2 (Modem 2 Signal): GPIO/24 active-low
  WLAN: GPIO/72 active-low
  WPS: GPIO/12 active-low
- Button:
  WPS / RESET: GPIO/34 active-low
- UART: 1x UART on PCB - 115200 8N1
- GPIO Watchdog: GPIO/62 mode=toggle timeout=1s
- PCIe: 2x miniPCIe for modem
- SIM Slots: 2x SIM Slots

Issue:
- No factory partition, eeprom is located
 at /lib/firmware/mt7620a.eeprom

Flash instruction:
Using UART:
1. Configure PC with a static IP address and setup an TFTP server.
2. Put rootfs into the tftp directory.
3. Connect the UART line as described on the PCB.
4. Power up the device and press Ctrl+C to break auto boot.
5. Use `system 6` command and follow the instruction to set device
   and tftp server IP address and input the rootfs file name.
   U-boot will then load the rootfs and write it into
   the flash.
6. Use `system 1` command and follow the instruction to set device
   and tftp server IP address and input the firmware file name.
   U-boot will then load the firmware once.
7. Login to LuCI and use LuCI upgrade firmware.

Original Firmware Dump / More details:
https://blog.gov.cooking/archives/research-hongdian-h8922-and-flash.html

Signed-off-by: Coia Prant <coiaprant@gmail.com>
Tested-by: Coia Prant <coiaprant@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17472
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

neheb commented on target/linux/ramips/dts/mt7620a_hongdian_h8922.dts:
```diff
@@ -0,0 +1,169 @@
+
+&wmac {
+	ralink,eeprom = "mt7620a.eeprom";
+};
```
newline please

rbqvq replied:
OK

neheb commented on target/linux/ramips/dts/mt7620a_hongdian_h8922.dts:
```diff
@@ -0,0 +1,169 @@
+};
+
+&wmac {
+	ralink,eeprom = "mt7620a.eeprom";
```
this is already default.

rbqvq replied:
OK

hauke commented on target/linux/ramips/mt7620/base-files/lib/firmware/soc_wmac.eeprom:
Is this binary file here really needed?

If it is also somewhere on the flash partition the driver should extract it dynamically or some script should do it in user space.

rbqvq replied:
Yes, it is necessary.

Most ramips devices have a factory partition for EEPROM.


But the EEPROM for this device is stored in /lib/firmware (original firmware)

And set ethernet mac from system.info file.

Wireless will not boot without EEPROM.

rbqvq replied:
In the original firmware, the device information SN and MAC are stored on the rootfs_data overlayfs

rbqvq replied:
This design is really terrible...

hauke replied:
Adding this to the base files will add it to all mt7620 images.

I think this should go into a separate firmware package which is then only installed on this board. Or into the wifi firmware package for this wifi chip.

rbqvq replied:
Yes, but this can be a fallback. If the EEPROM of the other board is damaged, the driver should try to load the firmware

rbqvq replied:
> I wonder if all of these devices share the same calibration data, ideally each device should have it's own.
> 
> You can use the `mediatek,eeprom-data` property to embed the calibration data, see [mt7986a-bananapi-bpi-r3-mini.dts](https://github.com/openwrt/openwrt/blob/04570f5ee2dfe6609c22d2d21678b14cd3314309/target/linux/mediatek/dts/mt7986a-bananapi-bpi-r3-mini.dts#L587) for example.

Thanks for your answer again, I checked rt2800 not supported `mediatek,eeprom-data`, I make a patch for that.
After I tested I will update this pull request.

rbqvq replied:
Tested passed, PR Updated.

hauke commented:
The instruction says:
> 2. Split the rootfs from firmware and put it into the tftp directory.
> It is recommended to use binwalk and dd commands.

If you need a different format for initial factory flashing please provide the images in the build process. 

Maybe this is working
```
  IMAGES += kernel.bin rootfs.bin
  IMAGE/kernel.bin := append-kernel
  IMAGE/rootfs.bin := append-rootfs  | pad-rootfs
```

rbqvq commented:
> The instruction says:
> 
> > 2. Split the rootfs from firmware and put it into the tftp directory.
> >    It is recommended to use binwalk and dd commands.
> 
> If you need a different format for initial factory flashing please provide the images in the build process.
> 
> Maybe this is working
> 
> ```
>   IMAGES += kernel.bin rootfs.bin
>   IMAGE/kernel.bin := append-kernel
>   IMAGE/rootfs.bin := append-rootfs  | pad-rootfs
> ```

Yes, rootfs is working.
Unfortunately, the kernel partition is only 1536k.
So we can't flash the kernel through uboot.

After flushing rootfs, boot through TFTP once, and then enter openwrt to flush the new full image.

rbqvq commented:
Dump original firmware here.
[HS8922.zip](https://github.com/user-attachments/files/18747765/HS8922.zip)

Flash layout
```
root@router:/sys/class/leds# cat /proc/mtd
dev:    size   erasesize  name
mtd0: 00030000 00010000 "u-boot"
mtd1: 00180000 00010000 "kernel"
mtd2: 00a50000 00010000 "rootfs"
mtd3: 00400000 00010000 "rootfs_data"
```
uboot (mtd0) include uboot and uboot-env, its in this PR.

No factory part found

rbqvq commented:
wow, thank you
I will try this method

Tianling Shen ***@***.***> 于 2025年2月11日周二 17:31写道：

> ***@***.**** commented on this pull request.
> ------------------------------
>
> On target/linux/ramips/mt7620/base-files/lib/firmware/soc_wmac.eeprom
> <https://github.com/openwrt/openwrt/pull/17472#discussion_r1950505362>:
>
> I wonder if all of these devices share the same calibration data, ideally
> each device should have it's own.
>
> You can use the mediatek,eeprom-data property to embed the calibration
> data, see mt7986a-bananapi-bpi-r3-mini.dts
> <https://github.com/openwrt/openwrt/blob/04570f5ee2dfe6609c22d2d21678b14cd3314309/target/linux/mediatek/dts/mt7986a-bananapi-bpi-r3-mini.dts#L587>
> for example.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/17472#discussion_r1950505362>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BMOQK5JE47OHE3HJDE3WKXT2PG7ORAVCNFSM6AAAAABUSKDE5WVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMBYGE2DONJXG4>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

hauke commented on target/linux/ramips/dts/mt7620a_hongdian_h8922-v30.dts:
```diff
@@ -0,0 +1,203 @@
+};
+
+&wmac {
+	ralink,eeprom-data = <0x20760501 0x000c4376 0x2058ffff 0xffffffff
```
Does this work?

There is a `mediatek,eeprom-data` attribute.

The indention is also wrong.

hauke commented on target/linux/ramips/dts/mt7620a_hongdian_h8922-v30.dts:
```diff
@@ -0,0 +1,203 @@
+};
+
+&wmac {
+	ralink,eeprom-data = <0x20760501 0x000c4376 0x2058ffff 0xffffffff
+						  0xffffffff 0xffffffff 0xffffffff 0xffffffff
+						  0xffffffff 0xffffffff 0x000c4376 0x2077000c
+						  0x43762066 0x220c0000 0xffffb701 0x5577a8aa
```
The indention is wrong here.

rbqvq replied:
It seems that GitHub and VSCode display tabs differently, and the length of tabs is different, so I use spaces instead, hoping there will be no problem

1715173329 replied:
please use tabs for indentation as always.

> It seems that GitHub and VSCode display tab differently

![image](https://github.com/user-attachments/assets/9c44e950-3f0d-4192-95af-c030658d81c2)

rbqvq replied:
Fixed

hauke commented on target/linux/ramips/dts/mt7620a_hongdian_h8922-v30.dts:
```diff
@@ -0,0 +1,203 @@
+#include "mt7620a.dtsi"
```
Please add a license header, for example:
```
// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
```

rbqvq replied:
Done

hauke commented:
Moving this to the device tree is fine. 
Please add the `603-wifi-rt2x00-Add-support-for-loading-EEPROM-from-devicetree-embedded-data.patch` patch to a separate commit.

rbqvq commented:
> Moving this to the device tree is fine.将其移动到设备树中是可以的。 Please add the `603-wifi-rt2x00-Add-support-for-loading-EEPROM-from-devicetree-embedded-data.patch` patch to a separate commit.请将 `603-wifi-rt2x00-Add-support-for-loading-EEPROM-from-devicetree-embedded-data.patch` 补丁添加到单独的提交中。

Ok, should I create a new pull request or just add a new commit here?

rbqvq committed:
mac80211: rt2x00: load the eeprom data from devicetree embedded data on Ralink SoCs

It will allow loading eeprom from eeprom-data embedded in device tree.
Ported from mediatek mt76 wireless driver (drivers/net/wireless/mediatek/mt76/eeprom.c)

Signed-off-by: Coia Prant <coiaprant@gmail.com>
Link: https://github.com/openwrt/openwrt/pull/17472
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

1715173329 commented on package/kernel/mac80211/patches/rt2x00/603-wifi-rt2x00-Add-support-for-loading-EEPROM-from-devicetree-embedded-data.patch:
```diff
@@ -0,0 +1,56 @@
+ .../net/wireless/ralink/rt2x00/rt2x00eeprom.c | 25 +++++++++++++++++++
+ 1 file changed, 25 insertions(+)
+
+diff --git a/drivers/net/wireless/ralink/rt2x00/rt2x00eeprom.c b/drivers/net/wireless/ralink/rt2x00/rt2x00eeprom.c
```
please refresh the patch with `make package/mt76/{clean,refresh}`

Cc @nbd168

rbqvq replied:
```shell
root@server /data/work/h8922/openwrt # make package/mt76/{clean,refresh}
make[2]: Entering directory '/data/work/h8922/openwrt/scripts/config'
make[2]: 'conf' is up to date.
make[2]: Leaving directory '/data/work/h8922/openwrt/scripts/config'
time: target/linux/prereq#0.76#0.52#1.27
 make[1] package/mt76/clean
 make[2] -C package/kernel/mt76 clean
make[2]: Entering directory '/data/work/h8922/openwrt/scripts/config'
make[2]: 'conf' is up to date.
make[2]: Leaving directory '/data/work/h8922/openwrt/scripts/config'
 make[1] package/mt76/refresh
 make[2] -C package/kernel/mt76 refresh
root@server /data/work/h8922/openwrt # git diff
root@server /data/work/h8922/openwrt # 
```
Done.
But nothing changed

1715173329 replied:
oh I'm stupid it should be `make package/mac80211/{clean,refresh}`

rbqvq replied:
```shell
-diff --git a/drivers/net/wireless/ralink/rt2x00/rt2x00eeprom.c b/drivers/net/wireless/ralink/rt2x00/rt2x00eeprom.c
-index 15c4e0b..aba3a43 100644
 --- a/drivers/net/wireless/ralink/rt2x00/rt2x00eeprom.c
 +++ b/drivers/net/wireless/ralink/rt2x00/rt2x00eeprom.c
-@@ -33,6 +33,27 @@ static void rt2800lib_eeprom_swap(struct rt2x00_dev *rt2x00dev)
+@@ -33,6 +33,27 @@ static void rt2800lib_eeprom_swap(struct
                rt2x00dev->eeprom[i] = swab16(rt2x00dev->eeprom[i]);
  }
  
@@ -43,7 +41,7 @@ index 15c4e0b..aba3a43 100644
  #if IS_ENABLED(CONFIG_MTD)
  static int rt2800lib_read_eeprom_mtd(struct rt2x00_dev *rt2x00dev)
  {
-@@ -193,6 +214,10 @@ int rt2x00lib_read_eeprom(struct rt2x00_dev *rt2x00dev)
+@@ -193,6 +214,10 @@ int rt2x00lib_read_eeprom(struct rt2x00_
  {
        int ret;
```

Is that right???

1715173329 replied:
Yep

rbqvq commented:
One more question, since this device has no factory partition, the default esw switch (eth0) mac is randomly generated by the kernel, should we generate 1 disposable mac to lan and wan in /etc/board.d/02_network (if the user keeps the config upgrade then the mac of the network card will not change), I would like to hear the community's suggestions on this

1715173329 commented:
> One more question, since this device has no factory partition, the default esw ethernet switch (eth0) mac is randomly generated by the kernel, should we generate 1 disposable mac to lan and wan in /etc/board.d/02_network (if the user keeps the config upgrade then the mac of the network card will not change)

Sounds good to me. This also avoids conflict when users have multiple devices running in the same network.

dangowrt commented:
Could we maybe inherit the LAN MAC address from the 4G modem...?
